core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -0,0 +1,974 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Minimal fixed point arithmetic primitives and types for runtime.
+
+#[cfg(feature = "std")]
+use crate::serde::{Serialize, Deserialize};
+
+use rstd::{
+	ops, prelude::*,
+	convert::{TryInto, TryFrom},
+	cmp::{min, max, Ordering}
+};
+use codec::{Encode, Decode};
+use crate::traits::{
+	SaturatedConversion, CheckedSub, CheckedAdd, Bounded, UniqueSaturatedInto, Saturating, Zero,
+};
+
+macro_rules! implement_per_thing {
+	($name:ident, $test_mod:ident, $max:tt, $type:ty, $upper_type:ty, $title:expr $(,)?) => {
+		/// A fixed point representation of a number between in the range [0, 1].
+		///
+		#[doc = $title]
+		#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug, Ord, PartialOrd))]
+		#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq)]
+		pub struct $name($type);
+
+		impl $name {
+			/// Nothing.
+			pub fn zero() -> Self { Self(0) }
+
+			/// `true` if this is nothing.
+			pub fn is_zero(&self) -> bool { self.0 == 0 }
+
+			/// Everything.
+			pub fn one() -> Self { Self($max) }
+
+			/// consume self and deconstruct into a raw numeric type.
+			pub fn deconstruct(self) -> $type { self.0 }
+
+			/// return the scale at which this per-thing is working.
+			pub const fn accuracy() -> $type { $max }
+
+			/// From an explicitly defined number of parts per maximum of the type.
+			///
+			/// This can be called at compile time.
+			pub const fn from_parts(parts: $type) -> Self {
+				Self([parts, $max][(parts > $max) as usize])
+			}
+
+			/// Converts from a percent. Equal to `x / 100`.
+			///
+			/// This can be created at compile time.
+			pub const fn from_percent(x: $type) -> Self {
+				Self([x, 100][(x > 100) as usize] * ($max / 100))
+			}
+
+			/// Return the product of multiplication of this value by itself.
+			pub fn square(self) -> Self {
+				// both can be safely casted and multiplied.
+				let p: $upper_type = self.0 as $upper_type * self.0 as $upper_type;
+				let q: $upper_type = $max as $upper_type * $max as $upper_type;
+				Self::from_rational_approximation(p, q)
+			}
+
+			/// Converts a fraction into `Permill`.
+			#[cfg(feature = "std")]
+			pub fn from_fraction(x: f64) -> Self { Self((x * ($max as f64)) as $type) }
+
+			/// Approximate the fraction `p/q` into a per-thing fraction. This will never overflow.
+			///
+			/// The computation of this approximation is performed in the generic type `N`. Given
+			/// `M` as the data type that can hold the maximum value of this per-thing (e.g. u32 for
+			/// perbill), this can only work if `N == M` or `N: From<M> + TryInto<M>`.
+			pub fn from_rational_approximation<N>(p: N, q: N) -> Self
+				where N: Clone + Ord + From<$type> + TryInto<$type> + ops::Div<N, Output=N>
+			{
+			// TODO: double check with @thiolliere the new bound. If your fraction type has a max
```

Comment:
@thiolliere I can really use a double-check on how I changed things here.

Reply:
oh I missed the transition from draft to reviewable I'll take a look for sure

Reply:
thanks 👍